transport/http2_client.go

```diff
@@ -542,6 +534,11 @@ func (t *http2Client) handleData(f *http2.DataFrame) {
 	data := make([]byte, size)
 	copy(data, f.Data())
 	s.write(recvMsg{data: data})
+
+	if f.FrameHeader.Flags.Has(http2.FlagDataEndStream) {
+		s.readClosed()
```

Comment:
s.statusCode and s.statusDesc need to be set properly to help user find out the server is broken (probably use codes.Internal for statusCode). The side effect if you set s.statusCode and statusDesc is that your refactoring of readClosed is ruined because I do not want grab and release s.mu twice here. I would suggest reverting your commit for readClosed and writeClosed for now and leave it as another PR when we figure out a good way to refactor.